clippy_utils/src/lib.rs

```diff
@@ -3370,3 +3370,24 @@ pub fn is_parent_stmt(cx: &LateContext<'_>, id: HirId) -> bool {
         Node::Stmt(..) | Node::Block(Block { stmts: &[], .. })
     )
 }
+
+/// Returns true if the given `expr` is a block or resembled as a block,
+/// such as `if`, `loop`, `match` expressions etc.
+pub fn is_block_like(expr: &Expr<'_>) -> bool {
+    matches!(
+        expr.kind,
+        ExprKind::Block(..) | ExprKind::ConstBlock(..) | ExprKind::If(..) | ExprKind::Loop(..) | ExprKind::Match(..)
+    )
+}
+
+/// Checks if the given `expr` is binary expression and any of its operands is "block like",
+/// see [`is_block_like`].
+///
+/// This could be useful to decide whether if a pair of parentheses should be used when offering
+/// suggestion.
+pub fn expr_contains_conjunctive_blocks(expr: &Expr<'_>) -> bool {
+    matches!(
+        expr.kind,
+        ExprKind::Binary(_, left, right) if is_block_like(left) || is_block_like(right)
```

Comment:
question: is this `|| is_block_like(right)` needed?

As far as I understand the issue, `if .. { .. } - 1` is parsed as `if .. { .. }; -1` (`-1` being its own unary expression), but this is only really an issue when the `if` is on the left side, isn't it?

This seems to work fine:
```rs
fn f() -> i32 {
  1 + if true { 2 } else { 3 }
}
```

Reply:
also, maybe we could eventually "copy" some of the logic that is in https://github.com/rust-lang/rust/blob/21e6de7eb64c09102de3f100420a09edc1a2a8d7/compiler/rustc_parse/src/parser/expr.rs#L387 , since that seems more accurate.
E.g. `if true { 2 } else { 3 } >> 4`  also does not need parentheses.

But maybe this is too complex for now. Just checking binary exprs like this seems fine for fixing the issue also

Reply:
> question: is this `|| is_block_like(right)` needed?

Well, I had that question at first, but then removing it breaks this specific test iirc:

https://github.com/rust-lang/rust-clippy/blob/d74f23a494205fba135f7095d7094e6abd65f335/tests/ui/needless_return.rs#L316

it kinda seems weird, I'll try test that again. Tbf, i was kinda struggle to find the perfect solution 😭, so I'll try that rustc_parse example you mentioned as well.

Reply:
Oh, doesn't this `expr_contains_conjunctive_blocks` function also need to be recursive? It currently only checks the immediate LHS and RHS, right?

Like `if {..} + 1 + 2` wouldn't return `true` since it stops at the `if {..} + 1` and doesn't look further into the LHS?

Reply:
oh yep, I forgot that 😂

Reply:
Ahh, sadly the `if true { 2 } else { 3 } >> 4` also fails to compile ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=fbb77867ee660d19c16825d616ede42b)), and neither does any [“can_continue_expr_unambiguously”](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_ast/util/parser/enum.AssocOp.html#method.can_continue_expr_unambiguously) operators.

Reply:
Huh... I swear I tried it and it worked. I must've done something wrong.🤔 Oh well, fine then. ^^